src/ejabberd_local.erl

```diff
@@ -127,18 +125,41 @@ process_iq(_, Acc, From, To, El) ->
 
 -spec process_iq_reply(From :: jid:jid(),
                        To :: jid:jid(),
-                       mongoose_acc:t(),
-                       IQ :: jlib:iq() ) -> mongoose_acc:t().
+                       Acc :: mongoose_acc:t(),
+                       IQ :: jlib:iq()) -> mongoose_acc:t().
 process_iq_reply(From, To, Acc, #iq{id = ID} = IQ) ->
-    case get_iq_callback(ID) of
-        {ok, undefined, Function} ->
-            Function(From, To, Acc, IQ);
-        {ok, Module, Function} ->
-            Module:Function(From, To, Acc, IQ);
+    case get_iq_callback_in_cluster(ID, Acc) of
+        {ok, Callback} ->
+            Callback(From, To, Acc, IQ);
         _ ->
             Acc
     end.
 
+-spec get_iq_callback_in_cluster(id(), mongoose_acc:t()) ->
+        {ok, callback()} | {error, term()}.
+get_iq_callback_in_cluster(ID, Acc) ->
+    {ok, NodeId} = ejabberd_node_id:node_id(),
+    BinNodeId = integer_to_binary(NodeId),
+    case binary:split(ID, <<"_">>, []) of
```

Comment:
I'd make this some kind of `parse_iq_id` next to `make_iq_id`, so that knowledge about the structured of this ID is encapsulated in two related and close-to-each-other functions, makes it a bit less magic.

But this still feels like a hack, so I'd say hacks are welcome but only if we really need one, still thinking about this one 🤔

Reply:
Could do `parse_iq_id` for sure.

yea, it is also used in pusher push to route a pubsub requests to a remote servers. D:
Which means that response could arrive from any s2s connection (meh).

So, the issue here is too generic API, which is now just writes data to other nodes which is not used :)

mod_caps/mod_ping - just local routes.
pusher_push - remote calls (s2s is possible).


BTW, ejabberd_local means jids like `servername` (without `user@` part).

That pusher_push code, will try to read it one more time:
```erlang
    ResponseHandler =
    fun(_From, _To, FAcc, Response) ->
            mod_event_pusher_push:cast(Host, fun handle_publish_response/5,
                                       [Host, To, PubsubJID, Node, Response]),
            FAcc
    end,
    %% The IQ is routed from the recipient's server JID to pubsub JID
    %% This is recommended in the XEP and also helps process replies to this IQ
    NotificationFrom = jid:make(<<>>, Host, <<>>),
    mod_event_pusher_push:cast(Host, fun ejabberd_local:route_iq/5,
                               [NotificationFrom, PubsubJID, Acc, Stanza, ResponseHandler]).
```